JSON serialization format (#2235)
open-telemetry/opentelemetry-specification: experimental/serialization/json.md

# JSON serialization

## Table of Contents

- [Overview](#overview)
- [Requirements](#requirements)
  - [Telemetry data requirements](#telemetry-data-requirements)
    - [Versioning](#versioning)
    - [Resource level](#resource-level)
    - [JSON encoding](#json-encoding)
  - [File storage requirements](#file-storage-requirements)
    - [JSON lines file](#json-lines-file)
    - [Streaming appending](#streaming-appending)
    - [Compression](#compression)
- [Example](#example)

## Overview

> This document describes the serialization of OpenTelemetry data as JSON objects, and their serialization to files.

Reviewer: ```suggestion
This document describes the serialization of OpenTelemetry data as JSON objects that can be stored in files.
```

## Requirements

### Telemetry data requirements

#### Versioning

> Each object serialized to JSON should be identified with a version tagging the version of the serializer.

Reviewer: Should this be the version of the serialization scheme?  Does the version of the serialization implementation matter?

Author: Yes, they're one and the same. Sorry for the confusion.

Author: I have attempted to clarify, please review.

> The deserializer must support and maintain compatibility with versions explicitly supported by this specification.

Reviewer: Does this specification define any versions of the serialization scheme?

Author: It's implicit and it's good you're asking this - it should be very explicit this spec defines version 1.

Author: I have attempted to clarify by creating a v1 header, please review

The version is serialized as a string under the key "version" of the JSON document.

#### Resource level

The JSON serialization operates at the level of `ResourceSpans`, `ResourceMetrics` and `ResourceLogs`.

> There is no support for serializing objects at a lower level.

Reviewer: There must be support for serializing types further down the hierarchy or these types could not be serialized.  Do you mean to say that these are the only valid types at the top level of a serialized JSON object that conforms to this spec?

Author: Yes indeed, that was brought up in discussions and I wanted to make it more clear.

Author: I have removed this sentence for now and clarified the top level resource objects.

Resources are serialized under the key "traces", "metrics" and "logs" respectively for `ResourceSpans`, `ResourceMetrics` and `ResourceLogs`.

#### JSON encoding

The data should be encoded according to the format specified in the [OTLP/HTTP standard](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).

### File storage requirements

#### JSON lines file

This file is a [JSON lines file](https://jsonlines.org/), and therefore follows those requirements:

* UTF-8 encoding
* Each line is a valid JSON value
* The line separator is `\n`
* The preferred file extension is `jsonl`.

#### Streaming appending

Data is appended sequentially to the file, irrespective of the type of data or timestamp.

This allows streaming data to the file.

#### Compression

> The file may be stored as uncompressed. Implementors are free to compress the file using gzip or tar compression algorithm as they see fit.

Reviewer: ```suggestion
The file may be stored as uncompressed. Implementors are free to compress the file as they see fit.
```

`tar` is not a compression algorithm and there's probably no need to specify which algorithms are usable if we're talking about files at rest.

## Table of Contents

- [Overview](#overview)
- [Requirements](#requirements)
  - [Telemetry data requirements](#telemetry-data-requirements)
    - [Versioning](#versioning)
    - [Resource level](#resource-level)
    - [JSON encoding](#json-encoding)
  - [File storage requirements](#file-storage-requirements)
    - [JSON lines file](#json-lines-file)
    - [Streaming appending](#streaming-appending)
    - [Compression](#compression)
- [Example](#example)

## Overview

This document describes the serialization of OpenTelemetry data as JSON objects that can be stored in files.

## Requirements

### Telemetry data requirements

#### Versioning

Each object serialized to JSON should be identified with a version tagging the version of the serialization scheme.

The deserializer must support and maintain compatibility with versions explicitly supported by this specification.

The version is serialized as a string under the key "version" of the JSON document.

### Serialization schema v1

This defines the first version of the serialization scheme.

#### Resource level

The JSON serialization accept top-level resource objects: `ResourceSpans`, `ResourceMetrics` and `ResourceLogs`.

> Resources are serialized under the key "traces", "metrics" and "logs" respectively for `ResourceSpans`, `ResourceMetrics` and `ResourceLogs`.

Reviewer 2: Perhaps instead of `traces`, `metrics`, `logs` those could be [`TracesData`](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L37), [`MetricsData`](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L37) and [`LogsData`](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto#L37) respectively, for keeping consistency with proto?

Author: OK, sure thing. Changed.

Reviewer 2: Great, thank you!

#### JSON encoding

The data should be encoded according to the format specified in the [OTLP/HTTP standard](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).

### File storage requirements

#### JSON lines file

This file is a [JSON lines file](https://jsonlines.org/), and therefore follows those requirements:

> * UTF-8 encoding

Author: It is unspecified.

* Each line is a valid JSON value
* The line separator is `\n`
* The preferred file extension is `jsonl`.

#### Streaming appending

Data is appended sequentially to the file, irrespective of the type of data or timestamp.

This allows streaming data to the file.

#### Compression

> The file may be stored as uncompressed. Implementors are free to compress the file as they see fit.

Reviewer 3: What does this mean? e.g. how would developer A consume the file from developer B if B chose to compress the file in an unknown format?

Author: Less is more - removing compression.

> {"version":"1.0.0.alpha","metrics":{"resourceMetrics":[{"resource":{"attributes":[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]},"instrumentationLibraryMetrics":[{"instrumentationLibrary":{},"metrics":[{"name":"counter-int","unit":"1","sum":{"dataPoints":[{"attributes":[{"key":"label-1","value":{"stringValue":"label-value-1"}}],"startTimeUnixNano":"1581452772000000321","timeUnixNano":"1581452773000000789","asInt":"123"},{"attributes":[{"key":"label-2","value":{"stringValue":"label-value-2"}}],"startTimeUnixNano":"1581452772000000321","timeUnixNano":"1581452773000000789","asInt":"456"}],"aggregationTemporality":"AGGREGATION_TEMPORALITY_CUMULATIVE","isMonotonic":true}},{"name":"counter-int","unit":"1","sum":{"dataPoints":[{"attributes":[{"key":"label-1","value":{"stringValue":"label-value-1"}}],"startTimeUnixNano":"1581452772000000321","timeUnixNano":"1581452773000000789","asInt":"123"},{"attributes":[{"key":"label-2","value":{"stringValue":"label-value-2"}}],"startTimeUnixNano":"1581452772000000321","timeUnixNano":"1581452773000000789","asInt":"456"}],"aggregationTemporality":"AGGREGATION_TEMPORALITY_CUMULATIVE","isMonotonic":true}}]}]}]}}

Reviewer 3: What does "atInt" mean?

Author: Sorry, I don't see a `atInt` in the code. Did you mean `asInt`?

Author: I am going to resolve this comment, `asInt` here is an element of the JSON serialization of the data point per the serialization framework set by OLTP/JSON as referred in the spec.

# JSON serialization

## Table of Contents

- [Overview](#overview)
- [Requirements](#requirements)
  - [Telemetry data requirements](#telemetry-data-requirements)
    - [Versioning](#versioning)
    - [Resource level](#resource-level)
    - [JSON encoding](#json-encoding)
  - [File storage requirements](#file-storage-requirements)
    - [UTF-8 preamble](#utf-8-preamble)
    - [File extension](#file-extension)
    - [Separation between objects](#separation-between-objects)
    - [Sequential append](#sequential-append)
    - [Compression](#compression)
- [Example](#example)

## Overview

> This document describes the serialization of OpenTelemetry data as JSON objects, and their serialization to files.

Author: I put some words towards that. Please let me know if that works.

#### JSON lines file

> This file is a [JSON lines file](https://jsonlines.org/), and therefore follows those requirements:

Reviewer 4: I am wondering if this spec would apply to "files" such as named pipes or stdout/stdin. If an application serializes OTLP in this format to stdout, it would need to be mixed signal.

Author: sure thing - how would you phrase it?

Reviewer 4: I have no issue with the phrasing in this section since it's specific to file storage. It would be good to clarify elsewhere if this intended as a general streaming serialization format in addition to file storage.

But I am worried we will need mixed signals in a single stream to accomplish shell pipelining or to capture a program's telemetry from stdout.

Author: I think I'd rather avoid a streaming discussion right now. Small and incremental scope, rather than trying to scope all of it out. If folks want streaming, there is OLTP support over HTTP and gRPC.

#### Versioning

Each object serialized to JSON should be identified with a version tagging the version of the serializer.

The deserializer must support and maintain compatibility with versions explicitly supported by this specification.

> The version is serialized as a string under the key "v" of the JSON document.

Reviewer 5: Given how verbose the rest of the keys in the format are should we use "version" for more clarity?

Author: ok with me

Data is appended sequentially to the file, irrespective of the timestamp.

This allows streaming data to the file.

### Telemetry data requirements

This defines the first version of the serialization scheme.

The data must be encoded according to the format specified in the [OTLP/HTTP standard](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).

> Only top-level objects, `ResourceSpans`, `ResourceMetrics`, and `ResourceLogs` are supported.

Reviewer 4: In the example below, the actual lines are `Export(Trace|Metrics|Logs)ServiceRequest` objects.

Author: I'm sorry, I'm confused and do not understand what you mean.

Reviewer 4: Sorry for the confusion. This json below is an [`ExportTraceServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/6a41aec384e8c79a1bf92d0105fc0f5a2f864f3f/opentelemetry/proto/collector/trace/v1/trace_service.proto#L35) message:

```json
{"resourceSpans":[{"resource":{"attributes":[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]},"instrumentationLibrarySpans":[{"instrumentationLibrary":{},"spans":[{"traceId":"","spanId":"","parentSpanId":"","name":"operationA","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","droppedAttributesCount":1,"events":[{"timeUnixNano":"1581452773000000123","name":"event-with-attr","attributes":[{"key":"span-event-attr","value":{"stringValue":"span-event-attr-val"}}],"droppedAttributesCount":2},{"timeUnixNano":"1581452773000000123","name":"event","droppedAttributesCount":2}],"droppedEventsCount":1,"status":{"deprecatedCode":"DEPRECATED_STATUS_CODE_UNKNOWN_ERROR","message":"status-cancelled","code":"STATUS_CODE_ERROR"}},{"traceId":"","spanId":"","parentSpanId":"","name":"operationB","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","links":[{"traceId":"","spanId":"","attributes":[{"key":"span-link-attr","value":{"stringValue":"span-link-attr-val"}}],"droppedAttributesCount":4},{"traceId":"","spanId":"","droppedAttributesCount":1}],"droppedLinksCount":3,"status":{}}]}]}]}
```

The [OTLP/HTTP spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp-request) mentioned in this PR for traces says:

> The request body is a Protobuf-encoded `ExportTraceServiceRequest` message.

In which case I think this line should say:
```suggestion
Only top-level objects, `ExportTraceServiceRequest`, `ExportMetricsServiceRequest `, and `ExportLogsServiceRequest` are supported.
```

Reviewer 4: If you want to divorce this spec from the RPC part of the protocol, there is also an equivalent [`TracesData`](https://github.com/open-telemetry/opentelemetry-proto/blob/6a41aec384e8c79a1bf92d0105fc0f5a2f864f3f/opentelemetry/proto/trace/v1/trace.proto#L37) message can you use here instead.

The data should be encoded according to the format specified in the [OTLP/HTTP standard](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).

### File storage requirements

#### UTF-8 preamble

The file must contain a byte order mark (BOM) of 3 bytes indicating that the file is encoded as UTF-8.

#### File extension

The file should use the `json` file extension.

#### Separation between objects

Each object should be serialized to JSON and stored to the file separated by a newline `\n` character.

#### Sequential append

Data is appended sequentially to the file, irrespective of the type of data or timestamp.

#### Compression

The file may be stored as uncompressed. Implementors are free to compress the file using gzip or tar compression algorithm as they see fit.

## Example

> ```json lines

Reviewer 5: Would it be useful to show the BOM here?

Reviewer 5: Can we have "rs", "rm" and "rl" in one json line or only one of the three is allowed?

Author: I thought showing the BOM might confuse folks since it's just 3 bytes.
Only one of the type of resources should be allowed.

> {"v":"1.0.0.alpha","rs":{"resourceSpans":[{"resource":{"attributes":[{"key":"resource-attr","value":{"stringValue":"resource-attr-val-1"}}]},"instrumentationLibrarySpans":[{"instrumentationLibrary":{},"spans":[{"traceId":"","spanId":"","parentSpanId":"","name":"operationA","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","droppedAttributesCount":1,"events":[{"timeUnixNano":"1581452773000000123","name":"event-with-attr","attributes":[{"key":"span-event-attr","value":{"stringValue":"span-event-attr-val"}}],"droppedAttributesCount":2},{"timeUnixNano":"1581452773000000123","name":"event","droppedAttributesCount":2}],"droppedEventsCount":1,"status":{"deprecatedCode":"DEPRECATED_STATUS_CODE_UNKNOWN_ERROR","message":"status-cancelled","code":"STATUS_CODE_ERROR"}},{"traceId":"","spanId":"","parentSpanId":"","name":"operationB","startTimeUnixNano":"1581452772000000321","endTimeUnixNano":"1581452773000000789","links":[{"traceId":"","spanId":"","attributes":[{"key":"span-link-attr","value":{"stringValue":"span-link-attr-val"}}],"droppedAttributesCount":4},{"traceId":"","spanId":"","droppedAttributesCount":4}],"droppedLinksCount":3,"status":{}}]}]}]}}

Reviewer 5: Is version number "1.0.0.alpha" coupled in any way to the OTLP Protobuf version number? What's the versioning strategy here? What are interoperability expectations between writers and readers? Should readers interpret the version in any way?

Author: The version number here is the version number of the serialization to json itself, irrespective and superseding the protobuf serialization. This means that it changes if the protobuf serialization changes, and it may also change if we change how we write the protobuf message. I would hope we'd keep tightly in line with the protobuf changes.

> # JSON serialization

Reviewer 5: This is specifically for file representation, not any JSON (we have OTLP/JSON which is different), so perhaps:
```suggestion
# JSON File serialization
```

> 

Reviewer 5: Add document status, similar to other spec docs.

Author: Sure thing, done.

## Table of Contents

- [Overview](#overview)
- [Requirements](#requirements)
  - [File storage requirements](#file-storage-requirements)
    - [JSON lines file](#json-lines-file)
    - [Streaming appending](#streaming-appending)
  - [Telemetry data requirements](#telemetry-data-requirements)
- [Examples](#examples)

## Overview

This document describes the serialization of OpenTelemetry data as JSON objects that can be stored in files.

## Requirements

### File storage requirements

#### JSON lines file

This file is a [JSON lines file](https://jsonlines.org/), and therefore follows those requirements:

* UTF-8 encoding
* Each line is a valid JSON value
* The line separator is `\n`
* The preferred file extension is `jsonl`.

#### Streaming appending

> Data is appended sequentially to the file, irrespective of the timestamp.

Reviewer 5: I think "appended sequentially" should not be a requirement. What if I have an implementation that can update a record in place?
If I understand correctly the intent here is that there is no guarantee that the timestamps in the file are monotonically increasing, so perhaps say that directly.

Reviewer 6: I think this is intending to say that timestamps do _not_ need to be monotonically increasing.

The question of whether the file must be append-only is interesting. Strictly speaking, the file _format_ does not necessarily need to be concerned with _how_ the file is written and read. However, I think it's worth considering that append-only files are going to be much more streamable, in the sense that the reader and writer can operate independently like a log file writer and reader. Without this requirement, tracking file rotation and updates becomes far more difficult.

Author: OK fixed.

This allows streaming data to the file.

### Telemetry data requirements

This defines the first version of the serialization scheme.

The data must be encoded according to the format specified in the [OTLP/HTTP standard](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).

Only top-level objects, `ExportTraceServiceRequest`, `ExportMetricsServiceRequest`, and `ExportLogsServiceRequest` are supported.

> Files must contain exactly one type of data: traces, metrics, or logs. Mixed content is not supported.

Reviewer 4: This line is stale, otherwise LGTM

Author: oh sorry. Fixing.
